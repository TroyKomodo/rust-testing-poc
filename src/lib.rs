#![cfg_attr(all(coverage_nightly, test), feature(coverage_attribute))]

pub fn add(a: i32, b: i32) -> i32 {
    a + b
}

pub fn subtract(a: i32, b: i32) -> i32 {
    a - b
}

pub fn multiply(a: i32, b: i32) -> i32 {
    a * b
}

pub fn divide(a: i32, b: i32) -> i32 {
    a / b
}

pub fn power(a: i32, b: u32) -> i32 {
    a.pow(b)
}

pub fn sqrt(a: f64) -> f64 {
    a.sqrt()
}

pub fn log(a: f64, base: f64) -> f64 {
    a.log(base)
}

pub fn test_branches(a: f64, b: f64) -> f64 {
    if a > b || a == b {
        sqrt(a) * sqrt(b) * 0.5
    } else {
        log(a, b)
    }
}

#[derive(Debug, PartialEq, serde_derive2::Serialize)]
pub enum TestEnum {
    A,
    B,
    B2,
    C,
}

pub fn test_branches_enum(a: TestEnum) -> TestEnum {
    if a == TestEnum::A {
        TestEnum::A
    } else if a == TestEnum::B || a == TestEnum::B2 {
        TestEnum::B
    } else {
        TestEnum::C
    }
}

pub fn bat_mans_chest(a: i32) -> i32 {
    panic!("IT ARIVED: {}", a);
}

fn big_brain_math(a: i32, mut b: i32) -> i32 {
    loop {
        for _ in 0..a {
            b += 1;
        }

        if b % 1027 == 0 {
            break;
        }
    }

    b
}

#[cfg_attr(all(coverage_nightly, test), coverage(off))]
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_add() {
        assert_eq!(add(1, 2), 3);
    }

    #[test]
    fn test_big_brain_math() {
        assert_eq!(big_brain_math(1, 2), 1027);
    }

    #[test]
    fn test_subtract() {
        assert_eq!(subtract(1, 2), -1);
        assert_eq!(subtract(2, 1), 1);
    }

    #[test]
    fn test_multiply() {
        assert_eq!(multiply(1, 2), 2);
    }

    #[test]
    fn test_divide() {
        assert_eq!(divide(1, 2), 0);
        assert_eq!(divide(2, 1), 2);
    }

    #[test]
    fn test_power() {
        assert_eq!(power(2, 3), 8);
    }

    #[test]
    fn test_sqrt() {
        assert_eq!(sqrt(4.0), 2.0);
    }

    #[test]
    fn test_log() {
        assert_eq!(log(10.0, 10.0), 1.0);
    }

    #[test]
    fn test_test_branches() {
        assert_eq!(test_branches(1.0, 2.0), 0.0);
        assert_eq!(test_branches(1.0, 1.0), 0.5);
        assert_eq!(test_branches(0.0, 0.0), 0.0);
    }

    #[test]
    fn test_snapshot() {
        insta::assert_compact_debug_snapshot!(test_branches(2.0, 1.0));
        insta::assert_compact_debug_snapshot!(test_branches(5.0, 1.2));
        insta::assert_compact_debug_snapshot!(test_branches(10.0, 1.0));
        insta::assert_compact_debug_snapshot!(test_branches(100.0, 1.0));
    }

    #[test]
    fn test_test_branches_enum() {
        assert_eq!(test_branches_enum(TestEnum::A), TestEnum::A);
        assert_eq!(test_branches_enum(TestEnum::B), TestEnum::B);
        assert_eq!(test_branches_enum(TestEnum::B2), TestEnum::B);
        assert_eq!(test_branches_enum(TestEnum::C), TestEnum::C);
    }

    #[test]
    #[should_panic]
    fn test_bat_mans_chest() {
        bat_mans_chest(1);
        assert_eq!(proc::magic_macro!(1, 2), 3);
    }

    #[test]
    fn test_derive() {
        insta::assert_snapshot!(postcompile::compile! {
            #[derive(Debug, Clone, serde_derive2::Deserialize)]
            struct Test {
                a: u32,
                b: i32,
            }

            const TEST: Test = Test { a: 1, b: proc::magic_macro2!(1, 2) };
        });
    }
}
