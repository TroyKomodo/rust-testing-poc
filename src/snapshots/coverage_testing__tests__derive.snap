---
source: src/lib.rs
expression: "postcompile::compile!\n{\n    #![allow(unused)] #[derive(Debug, Clone, serde_derive2::Deserialize)]\n    struct Test { a: u32, b: i32, } static TEST: Test = Test\n    { a: 1, b: proc::magic_macro2!(1, 2) }; fn test()\n    { let x = 1; let y = 2; let z = x + y; }\n}"
snapshot_kind: text
---
exit status: 0
--- stdout 
#![feature(prelude_import)]
#![allow(unused)]
#[prelude_import]
use std::prelude::rust_2021::*;
#[macro_use]
extern crate std;
struct Test {
    a: u32,
    b: i32,
}
#[automatically_derived]
impl ::core::fmt::Debug for Test {
    #[inline]
    fn fmt(&self, f: &mut ::core::fmt::Formatter) -> ::core::fmt::Result {
        ::core::fmt::Formatter::debug_struct_field2_finish(
            f,
            "Test",
            "a",
            &self.a,
            "b",
            &&self.b,
        )
    }
}
#[automatically_derived]
impl ::core::clone::Clone for Test {
    #[inline]
    fn clone(&self) -> Test {
        Test {
            a: ::core::clone::Clone::clone(&self.a),
            b: ::core::clone::Clone::clone(&self.b),
        }
    }
}
#[doc(hidden)]
#[allow(non_upper_case_globals, unused_attributes, unused_qualifications)]
const _: () = {
    #[allow(unused_extern_crates, clippy::useless_attribute)]
    extern crate serde as _serde;
    #[automatically_derived]
    impl<'de> _serde::Deserialize<'de> for Test {
        fn deserialize<__D>(
            __deserializer: __D,
        ) -> _serde::__private::Result<Self, __D::Error>
        where
            __D: _serde::Deserializer<'de>,
        {
            #[allow(non_camel_case_types)]
            #[doc(hidden)]
            enum __Field {
                __field0,
                __field1,
                __ignore,
            }
            #[doc(hidden)]
            struct __FieldVisitor;
            impl<'de> _serde::de::Visitor<'de> for __FieldVisitor {
                type Value = __Field;
                fn expecting(
                    &self,
                    __formatter: &mut _serde::__private::Formatter,
                ) -> _serde::__private::fmt::Result {
                    _serde::__private::Formatter::write_str(
                        __formatter,
                        "field identifier",
                    )
                }
                fn visit_u64<__E>(
                    self,
                    __value: u64,
                ) -> _serde::__private::Result<Self::Value, __E>
                where
                    __E: _serde::de::Error,
                {
                    match __value {
                        0u64 => _serde::__private::Ok(__Field::__field0),
                        1u64 => _serde::__private::Ok(__Field::__field1),
                        _ => _serde::__private::Ok(__Field::__ignore),
                    }
                }
                fn visit_str<__E>(
                    self,
                    __value: &str,
                ) -> _serde::__private::Result<Self::Value, __E>
                where
                    __E: _serde::de::Error,
                {
                    match __value {
                        "a" => _serde::__private::Ok(__Field::__field0),
                        "b" => _serde::__private::Ok(__Field::__field1),
                        _ => _serde::__private::Ok(__Field::__ignore),
                    }
                }
                fn visit_bytes<__E>(
                    self,
                    __value: &[u8],
                ) -> _serde::__private::Result<Self::Value, __E>
                where
                    __E: _serde::de::Error,
                {
                    match __value {
                        b"a" => _serde::__private::Ok(__Field::__field0),
                        b"b" => _serde::__private::Ok(__Field::__field1),
                        _ => _serde::__private::Ok(__Field::__ignore),
                    }
                }
            }
            impl<'de> _serde::Deserialize<'de> for __Field {
                #[inline]
                fn deserialize<__D>(
                    __deserializer: __D,
                ) -> _serde::__private::Result<Self, __D::Error>
                where
                    __D: _serde::Deserializer<'de>,
                {
                    _serde::Deserializer::deserialize_identifier(
                        __deserializer,
                        __FieldVisitor,
                    )
                }
            }
            #[doc(hidden)]
            struct __Visitor<'de> {
                marker: _serde::__private::PhantomData<Test>,
                lifetime: _serde::__private::PhantomData<&'de ()>,
            }
            impl<'de> _serde::de::Visitor<'de> for __Visitor<'de> {
                type Value = Test;
                fn expecting(
                    &self,
                    __formatter: &mut _serde::__private::Formatter,
                ) -> _serde::__private::fmt::Result {
                    _serde::__private::Formatter::write_str(__formatter, "struct Test")
                }
                #[inline]
                fn visit_seq<__A>(
                    self,
                    mut __seq: __A,
                ) -> _serde::__private::Result<Self::Value, __A::Error>
                where
                    __A: _serde::de::SeqAccess<'de>,
                {
                    let __field0 = match _serde::de::SeqAccess::next_element::<
                        u32,
                    >(&mut __seq)? {
                        _serde::__private::Some(__value) => __value,
                        _serde::__private::None => {
                            return _serde::__private::Err(
                                _serde::de::Error::invalid_length(
                                    0usize,
                                    &"struct Test with 2 elements",
                                ),
                            );
                        }
                    };
                    let __field1 = match _serde::de::SeqAccess::next_element::<
                        i32,
                    >(&mut __seq)? {
                        _serde::__private::Some(__value) => __value,
                        _serde::__private::None => {
                            return _serde::__private::Err(
                                _serde::de::Error::invalid_length(
                                    1usize,
                                    &"struct Test with 2 elements",
                                ),
                            );
                        }
                    };
                    _serde::__private::Ok(Test { a: __field0, b: __field1 })
                }
                #[inline]
                fn visit_map<__A>(
                    self,
                    mut __map: __A,
                ) -> _serde::__private::Result<Self::Value, __A::Error>
                where
                    __A: _serde::de::MapAccess<'de>,
                {
                    let mut __field0: _serde::__private::Option<u32> = _serde::__private::None;
                    let mut __field1: _serde::__private::Option<i32> = _serde::__private::None;
                    while let _serde::__private::Some(__key) = _serde::de::MapAccess::next_key::<
                        __Field,
                    >(&mut __map)? {
                        match __key {
                            __Field::__field0 => {
                                if _serde::__private::Option::is_some(&__field0) {
                                    return _serde::__private::Err(
                                        <__A::Error as _serde::de::Error>::duplicate_field("a"),
                                    );
                                }
                                __field0 = _serde::__private::Some(
                                    _serde::de::MapAccess::next_value::<u32>(&mut __map)?,
                                );
                            }
                            __Field::__field1 => {
                                if _serde::__private::Option::is_some(&__field1) {
                                    return _serde::__private::Err(
                                        <__A::Error as _serde::de::Error>::duplicate_field("b"),
                                    );
                                }
                                __field1 = _serde::__private::Some(
                                    _serde::de::MapAccess::next_value::<i32>(&mut __map)?,
                                );
                            }
                            _ => {
                                let _ = _serde::de::MapAccess::next_value::<
                                    _serde::de::IgnoredAny,
                                >(&mut __map)?;
                            }
                        }
                    }
                    let __field0 = match __field0 {
                        _serde::__private::Some(__field0) => __field0,
                        _serde::__private::None => {
                            _serde::__private::de::missing_field("a")?
                        }
                    };
                    let __field1 = match __field1 {
                        _serde::__private::Some(__field1) => __field1,
                        _serde::__private::None => {
                            _serde::__private::de::missing_field("b")?
                        }
                    };
                    _serde::__private::Ok(Test { a: __field0, b: __field1 })
                }
            }
            #[doc(hidden)]
            const FIELDS: &'static [&'static str] = &["a", "b"];
            _serde::Deserializer::deserialize_struct(
                __deserializer,
                "Test",
                FIELDS,
                __Visitor {
                    marker: _serde::__private::PhantomData::<Test>,
                    lifetime: _serde::__private::PhantomData,
                },
            )
        }
    }
};
static TEST: Test = Test { a: 1, b: 1 + 2 };
fn test() {
    let x = 1;
    let y = 2;
    let z = x + y;
}
